examples/postgres/appengine-standard-java8/src/main/java/com/example/appengine/postgresql/PostgreSqlServlet.java

```diff
@@ -0,0 +1,122 @@
+/**
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.appengine.postgresql;
+
+import com.google.common.base.Stopwatch;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+// [START example]
+@SuppressWarnings("serial")
+// With @WebServlet annotation the webapp/WEB-INF/web.xml is no longer required.
+@WebServlet(name = "PostgreSQL",
+    description = "PostgreSQL: Write low order IP address to PostgreSQL",
+    urlPatterns = "/postgresql")
+public class PostgreSqlServlet extends HttpServlet {
+
+  Connection conn;
+
+  @Override
+  public void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException,
+      ServletException {
+
+    final String createTableSql = "CREATE TABLE IF NOT EXISTS visits ( visit_id SERIAL NOT NULL, "
+        + "user_ip VARCHAR(46) NOT NULL, ts timestamp NOT NULL, "
+        + "PRIMARY KEY (visit_id) );";
+    final String createVisitSql = "INSERT INTO visits (user_ip, ts) VALUES (?, ?);";
+    final String selectSql = "SELECT user_ip, ts FROM visits ORDER BY ts DESC "
+        + "LIMIT 10;";
+
+    String path = req.getRequestURI();
+    if (path.startsWith("/favicon.ico")) {
+      return; // ignore the request for favicon.ico
+    }
+
+    PrintWriter out = resp.getWriter();
+    resp.setContentType("text/plain");
+
+    // store only the first two octets of a users ip address
```

Comment:
This seems unnecessary and overly complicates the code.

Reply:
IP addresses are PII in this instance.

Reply:
So? Collecting full IP addresses is standard operating procedure. Apache and Nginx for example do it by default.

Reply:
The guide for installing Apache doesn't suggest turning off logging IP addresses:
https://cloud.google.com/compute/docs/tutorials/basic-webserver-apache

Reply:
Apache doesn't display it.

Reply:
But they log it, so storing it here is no different.

Reply:
When you retrieve the IP address from the database, you could get the binary representation with getAddress() and then display some of the bytes. That would be much simpler and avoid the correctness issues.
https://docs.oracle.com/javase/7/docs/api/java/net/InetAddress.html